### PR TITLE
8329433: Reduce nmethod header size

### PR DESCRIPTION
**Pull request by vnkozlov**
This is part of changes which try to reduce size of `nmethod` and `codeblob` data vs code in CodeCache.
These changes reduced size of `nmethod` header from 288 to 232 bytes. From 304 to 248 in optimized VM:
```
Statistics for 1282 bytecoded nmethods for C2:
 total in heap = 5560352 (100%)
 header = 389728 (7.009053%)
```
vs
```
Statistics for 1322 bytecoded nmethods for C2:
 total in heap  = 8307120 (100%)
 header         = 327856 (3.946687%)
```

Several unneeded fields in `nmethod` and `CodeBlob` were removed. Some fields were changed from `int` to `int16_t` with added corresponding asserts to make sure their values are fit into 16 bits.

I did additional cleanup after recent `CompiledMethod` removal.

Tested tier1-7,stress,xcomp and performance testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329433](https://bugs.openjdk.org/browse/JDK-8329433): Reduce nmethod header size (**Enhancement** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Igor Veresov](https://openjdk.org/census#iveresov) (@veresov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18768/head:pull/18768` \
`$ git checkout pull/18768`

Update a local copy of the PR: \
`$ git checkout pull/18768` \
`$ git pull https://git.openjdk.org/jdk.git pull/18768/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18768`

View PR using the GUI difftool: \
`$ git pr show -t 18768`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18768.diff">https://git.openjdk.org/jdk/pull/18768.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18768#issuecomment-2052650661)

### CODE REVIEW

File: src/hotspot/share/code/nmethod.hpp [hunk 1]
```diff
@@ -200,18 +200,20 @@ class nmethod : public CodeBlob {
...
 
+  CompiledICData* _compiled_ic_data;
+
   // offsets for entry points
-  address _entry_point;                      // entry point with class check
-  address _verified_entry_point;             // entry point without class check
-  address _osr_entry_point;                  // entry point for on stack replacement
+  address  _osr_entry_point;       // entry point for on stack replacement
+  uint16_t _entry_offset;          // entry point with class check
+  uint16_t _verified_entry_offset; // entry point without class check
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:208]: Changed direct entry pointers (8 bytes each in 64-bit VM) to offsets (2 bytes) to `code_begin()`.

File: src/hotspot/share/code/nmethod.hpp [hunk 2]
```diff
@@ -221,38 +223,24 @@ class nmethod : public CodeBlob {
...
-  int _nul_chk_table_offset;
+  uint16_t _skipped_instructions_size;
+
+  // _oops_offset == _data_offset,  offset where embedded oop table begins (inside data)
+  uint16_t _metadata_offset; // embedded meta data table
+  uint16_t _dependencies_offset;
+  uint16_t _scopes_pcs_offset;
+  int      _scopes_data_offset;
+  int      _handler_table_offset;
+  int      _nul_chk_table_offset;
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:234]: Changed data sections offsets base from `header_begin()` to `data_begin()`.
Note, `ScopesPcs` and `ScopesData` data could be > 64Kb.

File: src/hotspot/share/code/nmethod.hpp [hunk 3]
```diff
@@ -221,38 +223,24 @@ class nmethod : public CodeBlob {
...
 #if INCLUDE_JVMCI
-  int _speculations_offset;
-  int _jvmci_data_offset;
+  int      _speculations_offset;
+  int      _jvmci_data_offset;
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:237]: "Wasted space" when Graal is not used. May be address in future changes.

File: src/hotspot/share/code/nmethod.hpp [hunk 4]
```diff
@@ -264,14 +252,19 @@ class nmethod : public CodeBlob {
...
+  // whether to restart collecting RTM locking abort statistic again.
+  RTMState _rtm_state;
+#endif
 
   // Local state used to keep track of whether unloading is happening or not
   volatile uint8_t _is_unloading_state;
 
-  // used by jvmti to track if an event has been posted for this nmethod.
-  bool _load_reported;
-
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:274]: Converted to bit mask.

File: src/hotspot/share/code/nmethod.hpp [hunk 5]
```diff
@@ -721,7 +720,6 @@ class nmethod : public CodeBlob {
   // Note: _exception_cache may be read and cleaned concurrently.
   ExceptionCache* exception_cache() const         { return _exception_cache; }
   ExceptionCache* exception_cache_acquire() const;
-  void set_exception_cache(ExceptionCache *ec)    { _exception_cache = ec; }
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:724]: Not used.

File: src/hotspot/share/code/nmethod.hpp [hunk 6]
```diff
@@ -799,7 +797,7 @@ class nmethod : public CodeBlob {
   void unlink();
 
   // Deallocate this nmethod - called by the GC
-  void purge(bool free_code_cache_data, bool unregister_nmethod) override;
+  void purge(bool unregister_nmethod);
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.hpp:800]: Only `unregister_nmethod` is used by code.

File: src/hotspot/share/memory/heap.hpp [hunk 7]
```diff
@@ -38,8 +38,8 @@ class HeapBlock {
 
  public:
   struct Header {
-    size_t  _length;                             // the length in segments
-    bool    _used;                               // Used bit
+    uint32_t  _length;                           // the length in segments
+    bool      _used;                             // Used bit
```
Comment (vnkozlov) [inline src/hotspot/share/memory/heap.hpp:42]: The `Header` size is aligned up 8 bytes. With `size_t length;` the aligned size will be 16 bytes.

`HeapBlock/Header` is used by CodeCache only which has 2Gb limit. The `length` counts number of segments which smallest size (`CodeCacheSegmentSize`) is 64 bytes.  32 bits is enough to cover it.

File: src/hotspot/share/runtime/frame.cpp [hunk 8]
```diff
@@ -974,8 +974,9 @@ void frame::oops_nmethod_do(OopClosure* f, NMethodClosure* cf, DerivedOopClosure
 
     // Preserve potential arguments for a callee. We handle this by dispatching
     // on the codeblob. For c2i, we do
-    if (reg_map->include_argument_oops()) {
-      _cb->preserve_callee_argument_oops(*this, reg_map, f);
+    if (reg_map->include_argument_oops() && _cb->is_nmethod()) {
+      // Only nmethod preserves outgoing arguments at call.
+      _cb->as_nmethod()->preserve_callee_argument_oops(*this, reg_map, f);
```
Comment (vnkozlov) [inline src/hotspot/share/runtime/frame.cpp:979]: Only `nmethod` preserves arguments oops.

File: src/hotspot/share/code/nmethod.cpp [hunk 9]
```diff
@@ -1462,6 +1483,7 @@ nmethod::nmethod(
       jvmci_nmethod_data()->copy(jvmci_data);
     }
 #endif
+    dependencies->copy_to(this);
```
Comment (vnkozlov) [inline src/hotspot/share/code/nmethod.cpp:1486]: Missed to revert back this change.

File: src/hotspot/share/code/codeBlob.cpp [hunk 10]
```diff
@@ -73,57 +73,32 @@ unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size) {
...
   _relocation_size(align_up(cb->total_relocation_size(), oopSize)),
-  _content_offset(CodeBlob::align_code_offset(_header_size + _relocation_size)),
+  _content_offset(CodeBlob::align_code_offset(header_size + _relocation_size)),
   _code_offset(_content_offset + cb->total_offset_of(cb->insts())),
-  _frame_complete_offset(frame_complete_offset),
   _data_offset(_content_offset + align_up(cb->total_content_size(), oopSize)),
   _frame_size(frame_size),
   S390_ONLY(_ctable_offset(0) COMMA)
+  _header_size((uint16_t)header_size),
+  _frame_complete_offset((int16_t)frame_complete_offset),
```
Comment (dean-long) [inline src/hotspot/share/code/codeBlob.cpp:88]: Rather than a raw cast, it would be better to use checked_cast here, or better yet, change the incoming parameter types to match the field type.  That way, if the caller is passing a constant, the compiler can check it at compile time.

Comment (vnkozlov) [reply src/hotspot/share/code/codeBlob.cpp:88]: Agree and will do. In all case `sizeof(_Class_)` is used for `header_size`.

File: src/hotspot/share/code/nmethod.hpp [hunk 11]
```diff
@@ -280,7 +277,9 @@ class nmethod : public CodeBlob {
           _has_method_handle_invokes:1,// Has this method MethodHandle invokes?
           _has_wide_vectors:1,         // Preserve wide vectors at safepoints
           _has_monitors:1,             // Fastpath monitor detection for continuations
-          _has_flushed_dependencies:1; // Used for maintenance of dependencies (under CodeCache_lock)
+          _has_flushed_dependencies:1, // Used for maintenance of dependencies (under CodeCache_lock)
+          _is_unlinked:1,              // mark during class unloading
+          _load_reported:1;            // used by jvmti to track if an event has been posted for this nmethod
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.hpp:282]: It seems like the type could be changed from uint8_t to bool.

File: src/hotspot/share/code/nmethod.hpp [hunk 12]
```diff
@@ -188,8 +188,23 @@ class nmethod : public CodeBlob {
...
+      // These are used for compiled synchronized native methods to
+      // locate the owner and stack slot for the BasicLock. They are
+      // needed because there is no debug information for compiled native
+      // wrappers and the oop maps are insufficient to allow
+      // frame::retrieve_receiver() to work. Currently they are expected
+      // to be byte offsets from the Java stack pointer for maximum code
+      // sharing between platforms. JVMTI's GetLocalInstance() uses these
+      // offsets to find the receiver for non-static native wrapper frames.
+      ByteSize _native_receiver_sp_offset;
+      ByteSize _native_basic_lock_sp_offset;
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.hpp:205]: Don't we need an assert in the accessor functions to make sure nmethod is native or not?

Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.hpp:205]: I thought about that but in both places where these accessors are called (`frame::get_native_monitor()` and `frame::get_native_receiver()`) there are such asserts already:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/frame.cpp#L1085

File: src/hotspot/share/code/nmethod.hpp (hunk 11 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.hpp:282]: Is there difference in generated code when you use bool instead of uint8_t?
I used uint8_t to easy change to uint16_t in a future if needed.

File: src/hotspot/share/code/nmethod.cpp [hunk 13]
```diff
@@ -1236,35 +1227,51 @@ nmethod::nmethod(
...
-    _handler_table_offset    = _dependencies_offset;
+    int skipped_insts_size   = code_buffer->total_skipped_instructions_size();
+#ifdef ASSERT
+    assert(((skipped_insts_size >> 16) == 0), "size is bigger than 64Kb: %d", skipped_insts_size);
+
+    // SECT_CONSTS is first in code buffer so the offset should be 0.
+    int consts_offset = code_buffer->total_offset_of(code_buffer->consts());
+    assert(consts_offset == 0, "const_offset: %d", consts_offset);
+#endif
+    _skipped_instructions_size = (uint16_t)skipped_insts_size;
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.cpp:1241]:
```suggestion
    _skipped_instructions_size = checked_cast<uint16_t>(code_buffer->total_skipped_instructions_size());
```

File: src/hotspot/share/code/nmethod.cpp [hunk 14]
```diff
@@ -1236,35 +1227,51 @@ nmethod::nmethod(
...
-    _stub_offset             = content_offset()      + code_buffer->total_offset_of(code_buffer->stubs());
-    _oops_offset             = data_offset();
-    _metadata_offset         = _oops_offset          + align_up(code_buffer->total_oop_size(), oopSize);
-    _scopes_data_offset      = _metadata_offset      + align_up(code_buffer->total_metadata_size(), wordSize);
-    _scopes_pcs_offset       = _scopes_data_offset;
-    _dependencies_offset     = _scopes_pcs_offset;
-    _handler_table_offset    = _dependencies_offset;
+    int skipped_insts_size   = code_buffer->total_skipped_instructions_size();
+#ifdef ASSERT
+    assert(((skipped_insts_size >> 16) == 0), "size is bigger than 64Kb: %d", skipped_insts_size);
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.cpp:1235]:
```suggestion
```
I think it's simpler just to use checked_cast below.

File: src/hotspot/share/code/nmethod.cpp [hunk 15]
```diff
@@ -1236,35 +1227,51 @@ nmethod::nmethod(
...
+
+    // SECT_CONSTS is first in code buffer so the offset should be 0.
+    int consts_offset = code_buffer->total_offset_of(code_buffer->consts());
+    assert(consts_offset == 0, "const_offset: %d", consts_offset);
+#endif
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.cpp:1240]:
```suggestion
```

File: src/hotspot/share/code/nmethod.cpp [hunk 16]
```diff
@@ -1426,26 +1434,31 @@ nmethod::nmethod(
...
+    int sum_size      = oops_size + metadata_size + deps_size;
+    assert((sum_size >> 16) == 0, "data size is bigger than 64Kb: %d", sum_size);
 
-    _scopes_pcs_offset       = _scopes_data_offset   + align_up(debug_info->data_size       (), oopSize);
-    _dependencies_offset     = _scopes_pcs_offset    + adjust_pcs_size(debug_info->pcs_size());
-    _handler_table_offset    = _dependencies_offset  + align_up((int)dependencies->size_in_bytes(), oopSize);
-    _nul_chk_table_offset    = _handler_table_offset + align_up(handler_table->size_in_bytes(), oopSize);
+    _metadata_offset      = (uint16_t)oops_size;
+    _dependencies_offset  = _metadata_offset      + (uint16_t)metadata_size;
+    _scopes_pcs_offset    = _dependencies_offset  + (uint16_t)deps_size;
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.cpp:1445]: Use checked_cast instead of raw casts.

File: src/hotspot/share/code/nmethod.cpp [hunk 17]
```diff
@@ -1426,26 +1434,31 @@ nmethod::nmethod(
...
-    _nmethod_end_offset      = _nul_chk_table_offset + align_up(nul_chk_table->size_in_bytes(), oopSize);
+    DEBUG_ONLY( int data_end_offset = _nul_chk_table_offset + align_up(nul_chk_table->size_in_bytes(), oopSize); )
 #endif
-    _entry_point             = code_begin()          + offsets->value(CodeOffsets::Entry);
-    _verified_entry_point    = code_begin()          + offsets->value(CodeOffsets::Verified_Entry);
-    _osr_entry_point         = code_begin()          + offsets->value(CodeOffsets::OSR_Entry);
-    _exception_cache         = nullptr;
+    assert((data_offset() + data_end_offset) <= nmethod_size, "wrong nmethod's size: %d < %d", nmethod_size, (data_offset() + data_end_offset));
+
+    _entry_offset          = (uint16_t)offsets->value(CodeOffsets::Entry);
```
Comment (dean-long) [inline src/hotspot/share/code/nmethod.cpp:1459]: Use checked_cast.

File: src/hotspot/share/memory/heap.hpp [hunk 18]
```diff
@@ -51,13 +51,16 @@ class HeapBlock {
 
  public:
   // Initialization
-  void initialize(size_t length)                 { _header._length = length; set_used(); }
+  void initialize(size_t length)                 { set_length(length); set_used(); }
   // Merging/splitting
-  void set_length(size_t length)                 { _header._length = length; }
+  void set_length(size_t length)                 {
+    LP64_ONLY( assert(((length >> 32) == 0), "sanity"); )
+    _header._length = (uint32_t)length;
```
Comment (dean-long) [inline src/hotspot/share/memory/heap.hpp:58]:
```suggestion
    _header._length = checked_cast<uint32_t>length;
```

File: src/hotspot/share/memory/heap.hpp [hunk 19]
```diff
@@ -51,13 +51,16 @@ class HeapBlock {
...
+  }
 
   // Accessors
   void* allocated_space() const                  { return (void*)(this + 1); }
-  size_t length() const                          { return _header._length; }
+  size_t length() const                          { return (size_t)_header._length; }
```
Comment (dean-long) [inline src/hotspot/share/memory/heap.hpp:63]: This cast looks unnecessary.

File: src/hotspot/share/code/nmethod.hpp (hunk 12 above)
Comment (dean-long) [reply src/hotspot/share/code/nmethod.hpp:205]: OK, but I'd rather see it in the accessors too.  Some users are checking for method()->is_native() and others are checking for is_osr_method(), so we need to make sure those are always mutually exclusive: method()->is_native() != is_osr_method().

File: src/hotspot/share/code/nmethod.hpp (hunk 11 above)
Comment (dean-long) [reply src/hotspot/share/code/nmethod.hpp:282]: I don't think uint8_t vs uint16_t matters, only if it is signed, unsigned, or bool.  So if we have more than 8 individual :1 fields, it will expand to a 2nd byte.

File: src/hotspot/share/code/nmethod.hpp (hunk 12 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.hpp:205]: We have separate `nmethod()` constructor for native method wrappers and it sets `_entry_bci = InvocationEntryBci;`. So it is impossible to have OSR native method wrapper.
But I agree with adding assert into accessors to catch accidental usage for not native methods.

File: src/hotspot/share/code/nmethod.cpp (hunk 14 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1235]: Agree

File: src/hotspot/share/code/nmethod.cpp (hunk 15 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1240]: I assume you are suggesting to remove `#ifdef ASSERT`.
Done.

File: src/hotspot/share/code/nmethod.cpp (hunk 13 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1241]: Done.

File: src/hotspot/share/code/nmethod.cpp [hunk 20]
```diff
@@ -1426,26 +1434,31 @@ nmethod::nmethod(
...
     }
 
-    _oops_offset             = data_offset();
-    _metadata_offset         = _oops_offset          + align_up(code_buffer->total_oop_size(), oopSize);
-    _scopes_data_offset      = _metadata_offset      + align_up(code_buffer->total_metadata_size(), wordSize);
+    int oops_size     = align_up(code_buffer->total_oop_size(), oopSize);
+    int metadata_size = align_up(code_buffer->total_metadata_size(), wordSize);
+    int deps_size     = align_up((int)dependencies->size_in_bytes(), oopSize);
+    int sum_size      = oops_size + metadata_size + deps_size;
+    assert((sum_size >> 16) == 0, "data size is bigger than 64Kb: %d", sum_size);
```
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1441]: Okay. But I will put above code under `#ifdef ASSERT` then.

File: src/hotspot/share/code/nmethod.cpp (hunk 16 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1445]: okay

File: src/hotspot/share/code/nmethod.cpp (hunk 17 above)
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1459]: done

File: src/hotspot/share/memory/heap.hpp (hunk 18 above)
Comment (vnkozlov) [reply src/hotspot/share/memory/heap.hpp:58]: Done.

File: src/hotspot/share/memory/heap.hpp (hunk 19 above)
Comment (vnkozlov) [reply src/hotspot/share/memory/heap.hpp:63]: Agree.

**Review by vnkozlov — COMMENTED** [review]
@dean-long, thank you for review. I applied all your suggestions and push it after testing.

File: src/hotspot/share/code/codeBlob.cpp [hunk 21]
```diff
@@ -73,82 +73,58 @@ unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size) {
...
+  // probably wrong for tiered
+  assert(_frame_size >= -1, "must use frame size or -1 for runtime stubs");
+#endif // COMPILER1
 
   set_oop_maps(oop_maps);
 }
 
 // Simple CodeBlob used for simple BufferBlob.
-CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, int size, int header_size) :
+CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, int size, uint16_t header_size) :
```
Comment (coleenp) [inline src/hotspot/share/code/codeBlob.cpp:106]: Just a drive-by comment.  You might be able to use delegating constructors for CodeBlob so you don't have to have the field initializations twice.  Maybe the same for nmethod ?

Comment (vnkozlov) [reply src/hotspot/share/code/codeBlob.cpp:106]: Thank you, @coleenp, foe looking on these changes.

Which fields are initialized twice? Only `_oop_maps` is set to `nullptr` before we proper build oop maps in first constructor.

The only saving could be lines of code but then I would have to check that `cb != nullptr` and do other additional checks which I don't think will save much lines.

Separation of `nmethod` constructor for native wrappers is helping clear see the difference and I would like to keep them separate. We have `init_defaults()` method for similar code and I can move more code into it from both constructors.

Comment (coleenp) [reply src/hotspot/share/code/codeBlob.cpp:106]: Delegating constructors are the answer to having some common 'init' functions.  It would simply save lines of code that look the same in both constructor initializer lists.  But it's a drive-by comment.

Comment (vnkozlov) [reply src/hotspot/share/code/codeBlob.cpp:106]: Okay.

Comment (vnkozlov) [reply src/hotspot/share/code/codeBlob.cpp:106]: It is tempting to do for `nmethod` to replace `init_defaults()`. I will look what can be done.

Comment (vnkozlov) [reply src/hotspot/share/code/codeBlob.cpp:106]: It does not work. It does not allow fields initialization after delegation:
```
src/hotspot/share/code/nmethod.cpp: In constructor 'nmethod::nmethod(Method*, CompilerType, int, int, CodeOffsets*, CodeBuffer*, int, ByteSize, ByteSize, OopMapSet*)':
src/hotspot/share/code/nmethod.cpp:1232:56: error: mem-initializer for 'nmethod::<unnamed union>::<unnamed struct>::_native_receiver_sp_offset' follows constructor delegation
```

Yes, I can move some fields initialization into body - but then it will be duplicated initialization.
From reading net, the delegated constructor should take the largest number of arguments which default values are set by delegating constructors. It is not easy applicable to `nmethod` constructors - it will add more lines of code than remove. `nmethod` constructors we have are too big and different enough that delegation will not improve code.

I will keep `init_defaults()` and extend it to reduce common code in constructors.

File: src/hotspot/share/code/nmethod.hpp [hunk 22]
```diff
@@ -221,57 +238,37 @@ class nmethod : public CodeBlob {
...
   // location in frame (offset for sp) that deopt can store the original
   // pc during a deopt.
   int _orig_pc_offset;
 
-  int _compile_id;                        // which compilation made this nmethod
-
-  CompilerType _compiler_type;            // which compiler made this nmethod (u1)
-
-  bool _is_unlinked;
+  int          _compile_id;            // which compilation made this nmethod
```
Comment (JohnTortugo) [inline src/hotspot/share/code/nmethod.hpp:259]: NIT: are these fields always needed?

Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.hpp:259]: Yes, they are needed for debugging issues. They are important for error reporting, logs and events recording.  And they do not take much space: CompLevel and CompilerType are one byte size.

File: src/hotspot/share/code/nmethod.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (dean-long) [reply src/hotspot/share/code/nmethod.cpp:1439]: The ASSERT block above looks unnecessary, now that field assignments below are using checked_cast.

[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/code/nmethod.cpp:1439]: Agree, but I need to change how I use checked_cast below to get the same check as above. I will do it in next update:
```
 _dependencies_offset  = _metadata_offset      + checked_cast<uint16_t>(align_up(code_buffer->total_metadata_size(), wordSize));
 _scopes_pcs_offset    = _dependencies_offset  + checked_cast<uint16_t>(align_up((int)dependencies->size_in_bytes(), oopSize));
---
 _dependencies_offset  = checked_cast<uint16_t>(_metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize));
 _scopes_pcs_offset    = checked_cast<uint16_t>(_dependencies_offset  + align_up((int)dependencies->size_in_bytes(), oopSize));
```

**Comment by vnkozlov** [comment]
Merge [#18637](https://github.com/openjdk/jdk/pull/18637) added an other `short` field `_num_stack_arg_slots` which pushed `nmethod` size back to 240 bytes in product VM. I will not do changes in **this** PR to compensate it.

**Comment by vnkozlov** [comment]
I remove `ASSERT` blocks to address the last @dean-long comment.

**Review by dean-long — APPROVED** [review]

**Review by veresov — APPROVED** [review]
Looks good.

**Comment by vnkozlov** [comment]
Thank you @dean-long, @coleenp, @JohnTortugo  and @veresov for reviews.
